Feature/add science testing

dgergel opened the pull request:
This PR includes the following: 

1) adding point simulations with SnoTel and Ameriflux forcings for the VIC 5.0 testing infrastructure - compares current version of VIC 5 to previous versions (VIC 5.0 and VIC 4.2 - set up to compare future release candidates and versions with the initial VIC 5.0 release). SnoTel and Ameriflux simulations for VIC 4.2 and 5.0 are archived and for now are housed in the `VIC_test_data` Dropbox repo. This directory can be specified in `run_tests.py` with a new `arg.parser` optional argument (default is the `VIC_sample_data` directory). 
2) analysis plots to compare the point simulations. Includes 1) daily plots for SnoTel stations and 2) annual mean diurnal cycle and monthly mean diurnal cycle for Ameriflux sites

Changes to the following files: 

- run_tests.py

- test_utils.py - this now includes plotting functions for snotel and Ameriflux analysis as well as a string parsing function for getting lat/lon info from the Ameriflux site files 

- science.cfg 

- global_param.classic.4.2_snotel.txt

- global_param.classic.5.0_snotel.txt

- global_param.classic.4.2_ecflux.txt

- global_param.classic.5.0_ecflux.txt

jhamman commented on tests/run_tests.py:
```diff
@@ -22,6 +22,9 @@
+                        check_multistream,
+                        plot_science_tests)
+from test_image_driver import test_image_driver_no_output_file_nans,
                         check_multistream_classic)
```
It looks like there is a syntax error here.

jhamman commented on tests/run_tests.py:
```diff
@@ -139,14 +142,24 @@ def main():
+    parser.add_argument('--science_test_data_dir', type=str,
+                        help='directory to find science test data',
+                        default='./test_runs')
                         default=os.path.join(test_dir, '../samples/VIC_sample_data'))
```
another syntax error.

jhamman commented on tests/run_tests.py:
```diff
@@ -538,7 +556,8 @@ def run_science(config_file, vic_exe, test_data_dir, out_dir, driver):
         # read template global parameter file
-        infile = os.path.join(test_dir, test_dict['global_parameter_file'])
+        infile = os.path.join(test_dir, 'science', test_dict['global_parameter_file'])
+        print(infile)
```
remove print statement

jhamman commented on tests/science/global_param.classic.4.2_ecflux.txt:
```diff
@@ -0,0 +1,106 @@
+ALMA_INPUT  FALSE
+
+SOIL            $test_data_dir/datasets/ec_flux_towers/params/soil_param.site_test.txt
+ARC_SOIL        FALSE
```
remove this line.

jhamman commented on tests/science/global_param.classic.4.2_ecflux.txt:
```diff
@@ -0,0 +1,106 @@
+
+VEGLIB          $test_data_dir/datasets/ec_flux_towers/params/veg_lib_IGBP.fv
+VEGLIB_VEGCOVER TRUE
+# VEGLIB_IRR      FALSE
```
remove this line

jhamman commented on tests/science/global_param.classic.4.2_ecflux.txt:
```diff
@@ -0,0 +1,106 @@
+OUTVAR          OUT_LATENT_SUB          %.4f    *       *
+OUTVAR          OUT_R_NET               %.4f    *       *
+OUTVAR          OUT_SENSIBLE            %.4f    *       *
+OUTVAR          OUT_RAD_TEMP            %.4f    *       *
```
suggestion: change format from `%.4f` to `%.6g` or something with a bit more precision.

jhamman commented on tests/science/global_param.classic.4.2_snotel.txt:
```diff
@@ -0,0 +1,105 @@
+            # GF_406 = use (flawed) formulas for ground flux, deltaH, and fusion from VIC 4.0.6 and earlier;
+            # GF_410 = use formulas from VIC 4.1.0 (ground flux is correct, but deltaH and fusion ignore surf_atten);
+            # GF_FULL = use correct ground flux formula from VIC 4.1.0 and also take surf_atten into account in deltaH and fusion
+            # Default   = GF_FULL
```
remove documentation (comments) on options

jhamman commented on tests/science/global_param.classic.4.2_snotel.txt:
```diff
@@ -0,0 +1,105 @@
+
+SOIL            $test_data_dir/datasets/snotel/parameters/Snotel_VIC.4.1.2_soil.txt
+BASEFLOW    ARNO
+ARC_SOIL        FALSE
```
remove this line

jhamman commented on tests/science/global_param.classic.4.2_snotel.txt:
```diff
@@ -0,0 +1,105 @@
+OUTVAR          OUT_LATENT_SUB          %.4f    *       *
+OUTVAR          OUT_R_NET               %.4f    *       *
+OUTVAR          OUT_SENSIBLE            %.4f    *       *
+OUTVAR          OUT_RAD_TEMP            %.4f    *       *
```
see comment above on ASCII precision.

jhamman commented on tests/science/global_param.classic.5.0_ecflux.txt:
```diff
@@ -0,0 +1,105 @@
+OUTVAR          OUT_LATENT_SUB          %.4f    *       *
+OUTVAR          OUT_R_NET               %.4f    *       *
+OUTVAR          OUT_SENSIBLE            %.4f    *       *
+OUTVAR          OUT_RAD_TEMP            %.4f    *       *
```
see comment above on ASCII precision.

jhamman commented on tests/science/global_param.classic.5.0_ecflux.txt:
```diff
@@ -0,0 +1,105 @@
+OUTVAR          OUT_SENSIBLE            %.4f    *       *
+OUTVAR          OUT_RAD_TEMP            %.4f    *       *
+Status API Training Shop Blog About
+© 2016 GitHub, Inc. Terms Privacy Security Contact Help
```
remove this text...

jhamman commented on tests/science/global_param.classic.5.0_snotel.txt:
```diff
@@ -0,0 +1,90 @@
+OUTVAR          OUT_LATENT_SUB          %.4f    *       *
+OUTVAR          OUT_R_NET               %.4f    *       *
+OUTVAR          OUT_SENSIBLE            %.4f    *       *
+OUTVAR          OUT_RAD_TEMP            %.4f    *       *
```
see comment above on ASCII precision.

jhamman commented on tests/science/global_param.classic.5.0_snotel.txt:
```diff
@@ -0,0 +1,90 @@
+
+N_OUTFILES  1
+
+OUTFILE         outfile 24
```
update the output options for VIC 5 -- see the documentation for current options.

dgergel replied:
Added `AGGFREQ` as an output option (`NDAYS` for Snotel, `NHOURS` for Ameriflux)

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+def tsplit(string, delimiters):
+    '''Behaves like str.split but supports multiple delimiters.
+
+    '''
```
remove blank line

jhamman commented on tests/science/science.cfg:
```diff
@@ -1 +1,21 @@
+test_description = snotel point simulations
+global_parameter_file = global_param.classic.5.0_snotel.txt
+expected_retval = 0
+check = not_implemented
```
at a minimum, we can check for no-nans.

dgergel replied:
The no-nans check is still failing, I'll address that in a follow up PR if that's cool with you.

jhamman replied:
why is it failing?

dgergel replied:
It's failing in the `read_vic_ascii` step. I'm working on fixing it now.

dgergel replied:
Alright - so the `complete` check was failing because the output files have different start and end times depending on the length of the forcing files (differing forcing time series lengths between Fluxnet sites --> different output file time series start and end dates). The `output_file_nans` check was failing due to something weird going in in the `read_vic_ascii` function, which is now fixed.

jhamman commented on tests/science/science.cfg:
```diff
@@ -1 +1,21 @@
+plots = water_year
+observations_path = snotel
+vic4.2.d = vic_4_2
+vic5.0.0 = vic_5_0
```
I'd like to rethink this. What do you think about a list of previous simulations that we will compare to. Something like:

```
[science_test_snotel]
...
observations_path = snotel
...
[[compare]]
[[[VIC.4.2.d]]]
version = 4.2
archive = archive/vic_4_2/
[[[VIC.5.0.0.rc1]]]
version = 5.0
archive = archive/vic_5_0_rc1/
```
This will allow us to arbitrarily add in additional compare tests without having to change `run_tests.py`.

dgergel replied:
I think that's a great idea. I'll refactor it so that we can easily add previous versions or even branches.

jhamman commented on tests/test_utils.py:
```diff
@@ -8,9 +7,13 @@
-                                  default_vic_valgrind_error_code)
+from tonic.models.vic.vic import VICRuntimeError
+from tonic.models.vic.vic import default_vic_valgrind_error_code
+## Note ## that importing this is causing an error so it's commented out
```
still?

dgergel replied:
Nope, this works now with the recent updates to `tonic`.

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+            for j, _substring in enumerate(substack):
+                stack.insert(i+j, _substring)
+
+    return stack
```
can you give me an example usage of this function?  I'm thinking it is more complicated than it needs to be.

dgergel replied:
I used this to parse lat/lng info from the Ameriflux site files: 

```
with open(site_csv_file) as file:
                second_line = list(file)[1]
str_split = tsplit(second_line, ('Latitude: ', 'Longitude: ', 'Elevation (masl): '))
lat = str_split[1].strip()
lng = str_split[2].strip()
```

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+    return stack
+
+def plot_science_tests(driver, testname, science_test_data_dir, result_dir, plot_dir, vic_42_dir, vic_50_dir,
+                        obs_dir, plots_to_make):
```
related to my comment above, I'd like to generalize this a bit further so we don't need arguments like `vic_42_dir` and `vic_50_dir`.

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+                                os.path.join(science_test_data_dir, 'datasets', obs_dir, 'obs'),
+                                plots_to_make)
+    else:
+        print("this has not yet been implemented in the VIC 5.0 science test suite")
```
go ahead and just raise an error here.

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+    Returns
+    ----------
+    '''
+    if testname == "science_test_snotel":
```
I think it would be better if we used `test_type` instead of `testname`.  I could image a case where we want to have multiple snotel science tests using different parameters that both need `plot_snotel_comparison()`.  This would require a minor addition to the science test configuration file.

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+
+                    # plt VIC 4.2
+                    axarr[i].plot(ebal_42[vic_vars[i]].groupby(ebal_42['HOUR']).mean(),
+                                'b', label='VIC 4.2', linewidth=lw)
```
why aren't you using the pandas plot methods?  What's wrong with:

```Python
bal_42[vic_vars[i]].groupby(ebal_42['HOUR']).mean().plot(ax=axarr[i], 'b', label='VIC 4.2', linewidth=lw)
```

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+                    ebal_50x['HOUR'] = (ebal_50x['SEC'] * (1/3600)).astype(int)
+
+                    axarr[i].plot(ebal_50x[vic_vars[i]].groupby(ebal_50x['HOUR']).mean(),
+                                'y', label='VIC 5.0.x', linewidth=lw)
```
There is a bunch of repeated code in these lines (626:648).  A more extensible approach would be something like this:

*Psudo code...*
```Python

datasets = '''dict of datasets, all formatted identically'''

for key, dset in datasets.items():
    if varname in dset:
        dset[varname].groupby(lambda x: x.hour).mean().plot(ax=axarr[i], 'b', label=key, linewidth=lw)
```

You'll need to make sure all the datasets are formatted identically before you come into the plot step.  This means:
- All have datetime indexes (this will save you from needing to use the `'SEC'` columns.
- All all aligned (maybe easiest to just put them in the same dataframe)
- All have the same variable names (you may need to use `rename` in some cases)

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+                plotname = '%s_%s.png' % (lat, lng)
+                os.makedirs(os.path.join(plot_dir, 'annual_mean'), exist_ok=True)
+                savepath = os.path.join(plot_dir, 'annual_mean', plotname)
+                plt.savefig(savepath, bbox_inches='tight')
```
specify the dpi as 150 or something higher than the default.

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+
+            ebal_50x['DATES'] = pd.to_datetime(ebal_50x.YEAR * 10000 +
+                                             ebal_50x.MONTH * 100 +
+                                             ebal_50x.DAY, format='%Y%m%d')
```
factor out all the data munging into separate functions by dataset type.  This would also be a good place to use `read_vic_ascii()` in tonic.

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+
+            # column names for DataFrame
+            # column names
+            names = ['Year', 'Month', 'Day', 'Hour', 'P', 'Tair', 'SWdown', 'LWdown', 'RH', 'Patm',
```
rename this variable to ecflux names so it is clear that this is just for the obs.

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+        alpha = 1.0
+
+        if not lat_issue:
+            if 'annual_mean_diurnal_cycle' in plots_to_make:
```
I'd rather see the annual cycle.  `groupby(lambda x: x.month).mean()`

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+                plt.close()
+
+
+def plot_fluxnet_comparison(driver, testname, result_dir, plot_dir, vic_42_dir, vic_50_dir,
```
this function is really long and is somewhat hard to follow.  I think you want to break it out into a few pieces.

1.  get obs locations
2.  read obs data
3.  loop over model datasets and read them
4.  append obs data so that you have an iterable of like shaped series (`datasets.keys() == ['Observations', 'VIC.4.2.d', 'VIC.5.0.0.rc1', 'VIC.5.0.0.rc2', 'VIC.5.0.0', 'VIC.5.1.0']`)
5.  do analysis and make plots
  - annual cycle
  - diurnal cycle
  - other?

jhamman commented on tests/test_utils.py:
```diff
@@ -303,3 +306,437 @@ def check_multistream_classic(fnames):
+    else:
+        print("this has not yet been implemented in the VIC 5.0 science test suite")
+
+def plot_snotel_comparison(driver, testname, result_dir, plot_dir, vic_42_dir, vic_50_dir,
```
I ended up reviewing the ecflux function first but all my comments on breaking this out into smaller pieces, apply here as well.

jhamman commented:
@dgergel - 

I just finished my first review. Beyond the comments above, please post some example plots here and on our `#plot-review` slack channel.

jhamman commented on tests/test_utils.py:
```diff
@@ -8,9 +7,13 @@
 import pandas as pd
+import glob
+import re
+import matplotlib.pyplot as plt
```
please add matplotlib to the required dependencies in `VIC/ci/requirements.yml`

dgergel commented:
@jhamman I think I've addressed all of your comments. The `no-nans` check is still failing, but I think it might be better if I address that in a follow-up PR. Just a heads up, I've not yet dealt with the lat/lng precision issue I mentioned, so I'm just raising a warning for now.

jhamman commented on ci/requirements.yml:
```diff
@@ -10,6 +10,8 @@ dependencies:
   - xarray
   - bottleneck
+  - matplotlib
+  - warnings
```
`warnings` is a python builtin, you don't need to install it. This is currently causing your test failure.

jhamman commented on tests/science/vic_parameters_42.txt:
```diff
@@ -0,0 +1,277 @@
+atom .###############################################################################
```
fix typo.

jhamman commented on tests/run_tests.py:
```diff
@@ -139,14 +141,23 @@ def main():
 
+    # check to make sure science test data directory exists
+    science_test_data_dir = args.science_test_data_dir
+    if not os.path.exists(science_test_data_dir):
```
this needs to be: 

```Python
if 'science' in args.test and not os.path.exists(science_test_data_dir):
    ...
```

luckymichael commented on tests/science/global_param.classic.4.2_ecflux.txt:
```diff
@@ -0,0 +1,104 @@
+NLAYER      3
```
looks good better would be even better to align columns.

luckymichael commented on tests/science/vic_parameters_42.txt:
```diff
@@ -0,0 +1,276 @@
+ # @section DESCRIPTION
```
I like this format better than `global_param.classic.5.0_xxxx.txt` because it has explanation and units of the variables. The user can use/change this file more easily.

jhamman replied:
can we rename this file `vic_parameters_42_compat.txt`?

jhamman replied:
this file needs to be removed.

jhamman replied:
in the future, use `git mv old_filename new_filename` to rename a file.

luckymichael commented on tests/test_utils.py:
```diff
@@ -306,3 +312,480 @@ def check_multistream_classic(fnames):
+        vic_42['HOUR'] = (vic_42['SEC'] * (1/3600)
+                                        ).astype(int)
+
+    vic_42.index = vic_42.DATES
```
Can `vic_42` be subdaily outputs? If so, is `vic_42.DATES` unique and can it be used as index?

jhamman replied:
We should make the index a proper datetime index including the subdaily timesteps.

dgergel replied:
Updated to include subdaily timesteps. The reason that the method employed in the `read_vic_ascii` function for subdaily timesteps wasn't working for me was a pandas version difference in the `pd.to_datetime` function between 0.18.0 and 0.18.1 I believe.

luckymichael commented on tests/test_utils.py:
```diff
@@ -306,3 +312,480 @@ def check_multistream_classic(fnames):
+
+    return(snotel_swe)
+
+def read_vic_42(lat, lng, science_test_data_dir, items):
```
be more specific as `read_vic_42_output`?

dgergel commented:
@jhamman this should be ready to merge. I've updated the plotting functionality with input from group members and made some updates to the structure based on the code review from @luckymichael. There are a few more things with the current setup that I want to update, such as the lat/lng precision issue that we'd discussed, but I think that can be changed in a future PR.

jhamman commented on tests/test_utils.py:
```diff
@@ -164,6 +167,7 @@ def test_classic_driver_all_complete(fnames):
     start = None
     end = None
     for fname in fnames:
+        print(fname)
```
remove print statements

jhamman commented on tests/test_utils.py:
```diff
@@ -172,6 +176,8 @@ def test_classic_driver_all_complete(fnames):
             start = df.index[0]
             end = df.index[-1]
+            print("start = %s" %str(start))
+            print("end = %s" %str(end))
```
remove print statements

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+def read_snotel_swe_obs(filename, science_test_data_dir, items):
+
+    ''' reads in Snotel SWE obs and returns DataFrame
```
remove extra blank line

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+    snotel_swe = pd.read_csv(filename_fullpath,
+                            skiprows=0,
+                            delim_whitespace=True,
+                            names=['YEAR', 'MONTH', 'DAY', 'OUT_SWE'])
```
indentation is off

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+    snotel_swe['DATES'] = pd.to_datetime(snotel_swe.YEAR * 10000 +
+                                        snotel_swe.MONTH * 100 +
+                                        snotel_swe.DAY,
+                                        format='%Y%m%d')
```
indentation is off

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+    snotel_swe.index = snotel_swe['DATES']
+
+    return(snotel_swe)
```
return is a statement, not a function.  the parentheses are not needed here.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+    elif 'snotel' in items['compare_to']:
+        vic_42_file = 'outfile_%s_%s' % (lat, lng)
+        vic_42_dir = os.path.join(science_test_data_dir, 'test_runs',
+                                items['archive'], 'snotel', 'results')
```
indentation here and above.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+    vic_42 = pd.read_csv(os.path.join(vic_42_dir, vic_42_file),
+                            sep='\t',
+                            skiprows=5)
```
indentation here.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                                items['archive'], 'snotel', 'results')
+
+    else:
+        raise ValueError("this option has not yet been implemented")
```
which option, how about 

```Python
raise ValueError("this option (%s) has not yet been implemented" % items['compare_to'])
```

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+    ''' reads output from VIC 4.2
+    '''
+
+    if 'ecflux' in items['compare_to']:
```
what is the type of `items['compare_to']`?  A list or a string?

dgergel replied:
It's a string. So this should be re-worked.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+        vic_42 = vic_42.rename(columns=lambda x: x.replace('OUT_NET_SHORT',
+                                'OUT_SWNET'))
+        vic_42 = vic_42.rename(columns=lambda x: x.replace('OUT_NET_LONG',
+                                'OUT_LWNET'))
```
indentation, indentation, indentation....

dgergel replied:
Using a linter now.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+    vic_42.index = vic_42.DATES
+
+    return(vic_42)
```
`return vic_42`

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+def read_vic_5_output(lat, lng, science_test_data_dir, items):
+
+    ''' read VIC 5.0.x output
```
remove blank line before doc string.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                                items['archive'], 'snotel', 'results')
+
+    else:
+        raise ValueError("this option has not yet been implemented")
```
see comment above on 4.2 version of this function

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+    vic_5.index = vic_5.DATES
+
+    return(vic_5)
```
see comments above on datetime index and return statement.

dgergel replied:
Updated to account for subdaily data as in the `read_vic_42_output` function.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                                plot_dir,
+                                plots_to_make)
+    else:
+        print("this has not been implemented in the VIC 5.0 science test suite")
```
raise error instead.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                            result_dir, plot_dir,
+                            plots_to_make):
+    ''' makes snotel figures
+
```
remove blank line

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+    plot_units = ['mm', 'fraction', 'fraction', 'mm', '%', 'degrees C',
+                    'mm', '$W/{m^2}$', '$W/{m^2}$', '$W/{m^2}$']
+    lw = 4.0
+    sns.set_style('whitegrid')
```
maybe just set this at the import, unless you are going to use different styles for different plots.

dgergel replied:
I think I prefer not to set it at the import so that we can use other styles for plots that we might add in the future.

jhamman replied:
that is also fine but that is not how this function works.  If you want to sandbox the style behavior, you'll you need to use the `with sns.set_style():` syntax for each figure:

http://web.stanford.edu/~mwaskom/software/seaborn/tutorial/aesthetics.html#temporarily-setting-figure-style

dgergel replied:
Ah cool. I didn't realize I could do that - updated to set the style for each figure.

dgergel replied:
Interesting - that way doesn't seem to work, I don't fully understand why, there are lots of SO questions with the same error message. It fails with 'AttributeError: __exit__`. 

However, setting it locally like this works: 
```with plt.rc_context(dict(sns.axes_style("whitegrid"),
                                    **sns.plotting_context("notebook"))):
```

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+            if key == "snotel":
+
+                data[key] = read_snotel_swe_obs(filename,
```
lots of extra whitespace, too much is just as hard to read as too little.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+            if 'water_year' in plots_to_make:
+
+                fig, ax = plt.subplots(figsize=(10,10))
```
pep 8, space after `,`

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+    if os.listdir(os.path.join(obs_dir, subdir)) != []:
+        return(True)
+    else:
+        return(False)
```
see above on return statement

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                plt.close()
+
+def check_site_files(obs_dir, subdir):
+    if os.listdir(os.path.join(obs_dir, subdir)) != []:
```
I think this whole function could be

```Python
return len(os.listdir(os.path.join(obs_dir, subdir))) > 0
```

dgergel replied:
Indeed.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                if 'snotel' in data.keys():
+                    df_join = data.pop('snotel').merge(next(iter(data.values())),
+                                                    how='outer', on='DATES')
+                    df.index = df_join['DATES']
```
is `df_join` only used here? Shouldn't line 560 have taken care of the merging of the timeseries?

dgergel replied:
Yep, this section wasn't necessary - Line 560 was accomplishing that.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                                        items)
+
+        # loop over variables to plot
+        for i, plot_variable in enumerate(plot_variables):
```
`i` is never used.

dgergel replied:
This was being used to index into the units list, but I've updated this to including the units in a dict of plot variables.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+    except IndexError:
+        site_csv_file = glob.glob(os.path.join(obs_dir, subdir, 'us*.csv'))[0]
+    with open(site_csv_file) as file:
+        second_line = list(file)[1]
```
`file` is a Python built in, better to use `f` in this case.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+    lat = str_split[1].strip()
+    lng = str_split[2].strip()
+
+    return(lat, lng)
```
see above on `return`

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                                 ecflux_df.MONTH * 100 +
+                                 ecflux_df.DAY,format='%Y%m%d')
+
+    ecflux_df.index = ecflux_df.DATES
```
same comment as above on the index

dgergel replied:
Updated as in `read_vic_42_output` and `read_vic_5_output`.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+    if ('HOUR' not in ecflux_df) and ('SEC' in ecflux_df):
+        # add hour column for groupby
+        ecflux_df['HOUR'] = (ecflux_df['SEC'] * (1/3600)
+                                        ).astype(int)
```
I don't think this is correct for pep8

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+            vic_vars = ['OUT_LATENT', 'OUT_SENSIBLE', 'OUT_SWNET', 'OUT_LWNET']
+            variable_names = ['LH', 'H',
+                                'SW NET',
+                                'LW NET']
```
I might suggest putting a dictionary at the top of this module that is a mapping from VIC variable names to the labels we want to use on plots.  Then it can be reused by other comparisons.

Also good to do for units.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+            lw = 4.0
+            fs = 15
+            dpi = 150
+            sns.set_style('whitegrid')
```
move to top of module

dgergel replied:
Setting this prior to each figure.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                    annual_mean = {}
+                    for (key, df) in data.items():
+                        annual_mean[key] = pd.DataFrame(df[vic_var].groupby(
+                                                        df['HOUR']).mean())
```
I'd rather get rid of the `HOUR` key and just operate using the index.

dgergel replied:
Agreed, updated.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+                    # calculate annual mean diurnal cycle for each DataFrame
+                    annual_mean = {}
+                    for (key, df) in data.items():
```
no need for the parentheses around `key, df`

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                                    color=compare_data_dict[key]['color'],
+                                    linestyle=compare_data_dict[key]
+                                    ['linestyle'],
+                                    zorder=compare_data_dict[key]['zorder'])
```
sytax errors here I think

dgergel replied:
Not sure what you were specifically referring to, but this works.

jhamman replied:
this should be: 

```Python
series.plot(linewidth=compare_data_dict[key]['linewidth'], ...)
```

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+                    axarr[i].legend(loc='upper left')
+                    axarr[i].set_ylabel('%s ($W/{m^2}$)' %variable_names[i],
+                                        size=fs)
```
see comment above on moving units up to the module header

dgergel replied:
Moved units to the variable dict.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+                        df = pd.DataFrame({key: d[vic_var] for key, d in
+                                        monthly_mean.items() if vic_var in d})
+
```
it seems like you should be doing the groupby prior to the `for j in range(12)` line (772).  This may need a little reordering.

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                                        ['color'],
+                                        linestyle=compare_data_dict[key]
+                                        ['linestyle'],
+                                        zorder=compare_data_dict[key]['zorder'])
```
syntax

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+
+                        if i == 0 and j == 11:
+                            axarr[i,j].legend(loc='center left',
+                                                bbox_to_anchor=(1, 0.5))
```
I think it would clearer if you just wrote (outside of the for loops you are in:
```axarr[0, -1].legend(loc='center left', bbox_to_anchor=(1, 0.5))```

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                            axarr[i,j].legend(loc='center left',
+                                                bbox_to_anchor=(1, 0.5))
+                        axarr[i,j].set_ylabel('%s \n ($W/{m^2}$)'
+                                            %variable_names[i],
```
pep8

jhamman commented on tests/test_utils.py:
```diff
@@ -306,3 +313,513 @@ def check_multistream_classic(fnames):
+                        if i == 0:
+                            axarr[i,j].set_title(months[j], size=fs)
+                        '''if vic_var == 'OUT_LATENT':
+                            axarr[i,j].set_ylim([-50,300])'''
```
don't use multiline strings as comments, use comments instead.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                        sep='\t')
+
+    # remove spaces from column names
+    vic_5 = vic_5.rename(columns=lambda x: x.replace(' ', ''))
```
since you are overwritting the orginal `vic_5`, better to just use the `inplace` keyword argument:

```Python
vic_5.rename(columns=lambda x: x.replace(' ', ''), inplace=True)
```

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+                        series.plot(use_index=True,
+                                    linewidth=compare_data_dict[key]
+                                    ['linewidth'],
```
I still want this to be fixed.  The `['linewidth'],` should go on the line above. The proper way to handle this is:

```Python
                    for key, series in df.iteritems():
                        series.plot(
                            use_index=True,
                            linewidth=compare_data_dict[key]['linewidth'],
                            ax=ax,
                            color=compare_data_dict[key]['color'],
                            linestyle=compare_data_dict[key]['linestyle'],
                            zorder=compare_data_dict[key]['zorder'])
```

dgergel replied:
Ah thanks for the example!

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+    if 'HOUR' in ecflux_df:
+        ecflux_df.index += pd.Series([pd.Timedelta(s, unit='s') for s in
+                                     ecflux_df['HOUR'] * 3600],
+                                     index=ecflux_df.index)
```
couldn't you just do 

```Python
    if 'HOUR' in ecflux_df:
        ecflux_df.index += pd.Series([pd.Timedelta(s, unit='h') for s in
                                     ecflux_df['HOUR']],
                                     index=ecflux_df.index)
```

dgergel replied:
Yep, that works.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+    '''if ('HOUR' not in ecflux_df) and ('SEC' in ecflux_df):
+        # add hour column for groupby
+        ecflux_df['HOUR'] = (ecflux_df['SEC'] * (1/3600)
+                                        ).astype(int)'''
```
don't use strings as a comment and don't commit commented out code.

dgergel replied:
Oops! Accidentally overlooked that.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                    plt.close()
+
+        else:
+            subdir_files = False
```
`subdir_files` is never used.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                                                 linestyle=compare_data_dict
+                                                 [key]['linestyle'],
+                                                 zorder=compare_data_dict[key]
+                                                 ['zorder'])
```
see comments above on line continuation.

dgergel replied:
I've updated this as you suggested before with a smaller hanging indent, but the line is still too short - the `linewidth=compare_data_dict[key]['linewidth']` line still has to be broken up.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                                                                  month,
+                                                                  df.index.
+                                                                  hour]).
+                                                                 mean())
```
this is pretty ugly.  How about: 

```Python
                                monthly_mean[key] = pd.DataFrame(
                                    df[vic_var].groupby(
                                        [df.index.month,
                                         df.index.hour]). mean())
```

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+            # make figures
+            vars = {'OUT_LATENT': 'LH', 'OUT_SENSIBLE': 'H', 'OUT_SWNET':
+                    'SW_NET', 'OUT_LWNET': 'LW NET'}
```
vars is a built in,  rename it and move this dict to the module header.

dgergel replied:
Done.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                                            sharex=True,
+                                            sharey=True)
+
+                    months = ['January', 'February', 'March', 'April', 'May',
```
move months to header since we don't need to define this inside the for loop.

dgergel replied:
Done.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                              'October', 'November', 'December']
+
+                    for i, (vic_var, variable_name) in enumerate(vars.items()):
+                        for j in range(12):
```
```Python
for j, month in enumerate(months):```

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                                              in d})
+
+                            for key, series in df.iteritems():
+                                series[j+1].plot(linewidth=compare_data_dict
```
nit: space around `+`

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+                            # calculate monthly mean diurnal cycle
+                            monthly_mean = {}
+                            for (key, df) in data.items():
```
I think the for loop over months (line 782) and `data.items()` (line 786) can be switched. The current implementation is doing the expensive groupby-apply-combine steps in the groupby 11 times more than it needs to.

dgergel replied:
Yep, updated.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                        for key, df in data.items():
+                            annual_mean[key] = pd.DataFrame(df[vic_var].
+                                                            groupby(df.index.
+                                                            hour).mean())
```
see comments my other comments on line breaking. Breaking after a `.` should be your last resort.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+                    except OSError:
+                        warnings.warn("this site has a " +
+                                      "lat/lng precision issue")
```
The plus sign (`+`) is not needed here.

This would be better syntax though:

```Python
                        warnings.warn(
                            "this site has a lat/lng precision issue")
```

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                            result_dir, plot_dir,
+                            plots_to_make):
+
+    ''' makes Ameriflux figures
```
you still have extra blank lines floating around. You don't need (or want) them between your function definition and your docstrings.

jhamman replied:
[pep8 says](https://www.python.org/dev/peps/pep-0008/#blank-lines):

> Extra blank lines may be used (sparingly) to separate groups of related functions. Blank lines may be omitted between a bunch of related one-liners (e.g. a set of dummy implementations).

>Use blank lines in functions, sparingly, to indicate logical sections.

dgergel replied:
Cool, thanks. I went a little overboard with adding blank lines for readability.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+             'OUT_SOIL_MOIST2', 'OUT_SOIL_MOIST3', 'OUT_SOIL_MOIST4',
+             'OUT_SOIL_MOIST5', 'OUT_SOIL_TEMP1', 'OUT_SOIL_TEMP2',
+             'OUT_SOIL_TEMP3', 'OUT_SOIL_TEMP4', 'OUT_SOIL_TEMP5', 'OUT_SWNET',
+             'OUT_LWNET', 'OUT_SENSIBLE', 'OUT_LATENT', 'OUT_GRND_FLUX']
```
move this to the header and call it `fluxnet_names`

dgergel replied:
Good idea, updated as you suggested.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+def read_vic_5_output(lat, lng, science_test_data_dir, items):
+    ''' Read VIC 5.0.x output.
+    '''
```
pep8 nit: the closing triple ticks should be on the same line

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+        vic_42['SEC'] = vic_42.HOUR * 3600
+        vic_42.index += pd.Series([pd.Timedelta(s, unit='s') for s in
+                                  vic_42['SEC']],
+                                  index=vic_42.index)
```
see my other comment on this.  Just use a timedelta with `unit=hour`

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+def read_vic_42_output(lat, lng, science_test_data_dir, items):
+    ''' Reads output from VIC 4.2.
+    '''
```
one line docstring

dgergel replied:
your docstrings aren't one line! I'm going to update them while I'm at it.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+def read_snotel_swe_obs(filename, science_test_data_dir, items):
+    '''Reads in Snotel SWE obs and returns DataFrame.
+    '''
```
one line docstring.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                    # save plot
+                    plotname = '%s_%s.png' % (lat, lng)
+                    print(lat)
+                    print(lng)
```
remove these print statements

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                                exist_ok=True)
+                    savepath = os.path.join(plot_dir, 'annual_mean', plotname)
+                    plt.savefig(savepath, bbox_inches='tight', dpi=dpi)
+                    print("saved figure to %s" % savepath)
```
remove print statement

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+                            df = pd.DataFrame({key: d[vic_var] for key, d in
+                                              monthly_mean.items() if vic_var
+                                              in d})
```
fix indentation

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+                            axarr[i, j].set_ylabel('%s \n ($W/{m^2}$)'
+                                                   % variable_name,
+                                                   size=fs)
```
break this line in a bit cleaner fashion:

```Python
                            axarr[i, j].set_ylabel(
                                '%s \n ($W/{m^2}$)' % variable_name, size=fs)
```

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                    savepath = os.path.join(plot_dir,
+                                            'monthly_mean', plotname)
+                    plt.savefig(savepath, bbox_inches='tight', dpi=dpi)
+                    print("saved figure to %s" % savepath)
```
remove print statement

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                                                     science_test_data_dir,
+                                                     items)
+                    except OSError:
+                        warnings.warn("this site does not have data")
```
which site?

dgergel replied:
Changed this to print the site name in the error message.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+        else:
+            subdir_files = False
+            warnings.warn("this site does not have data")
```
which site?

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                      'OUT_SNOW_CANOPY': '%', 'OUT_SNOW_PACK_TEMP':
+                      'degrees C', 'OUT_SNOW_MELT': 'mm', 'OUT_R_NET':
+                      '$W/{m^2}$', 'OUT_LATENT': '$W/{m^2}$',
+                      'OUT_SENSIBLE': '$W/{m^2}$'}
```
move this to the module header

dgergel replied:
But it's specific to this function?

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+    if 'SEC' in vic_5:
+        vic_5.index += pd.Series([pd.Timedelta(s, unit='s') for s in
+                                 vic_5['SEC']],
```
you are missing a single space in the indentation here.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+    # add datetime index
+    time_cols = ['YEAR', 'MONTH', 'DAY']
+    vic_5.index = pd.to_datetime(vic_5[time_cols])
```
this is really clean, why don't you do this in lines 210 and 354?

dgergel replied:
Because I forgot to update those lines. Doing that now.

jhamman commented on tests/test_utils.py:
```diff
@@ -203,9 +207,11 @@ def read_vic_ascii(filepath, parse_dates=True, datetime_index=None, sep='\t',
         time_cols = ['YEAR', 'MONTH', 'DAY']
-        df.index = pd.to_datetime(df[time_cols])
+        df.index = pd.to_datetime(df.YEAR * 10000 + df.MONTH * 100 +
+                                  df.DAY, format='%Y%m%d')
```
I think you should change this behavior back to the way it was before.

dgergel replied:
Done. This was because of the pandas version issue I was having between `0.18.0` and `0.18.1` with the behavior of `pd.to_datetime`.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+
+    vic_5 = pd.read_csv(os.path.join(vic_5_dir, vic_5_file),
+                        skiprows=2,
+                        sep='\t')
```
why aren't you using `read_vic_ascii()` here?

dgergel replied:
There were some aspects of the `read_vic_5_output` function that are specific to reading fluxnet and snotel data.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+                    fig, ax = plt.subplots(figsize=(10, 10))
+
+                    df = pd.DataFrame({key: d[plot_variable] for key, d in
+                                      data.items() if plot_variable in d})
```
indentation is off by one space.

jhamman commented on tests/test_utils.py:
```diff
@@ -298,11 +306,536 @@ def check_multistream_classic(fnames):
+            if 'water_year' in plots_to_make:
+
+                with plt.rc_context(dict(sns.axes_style("whitegrid"),
+                                    **sns.plotting_context("notebook"))):
```
this is fine but I think it would make more sense to just make a module level decision here. 

style: `whitegrid`
context: `paper`

dgergel replied:
See what you think of the way I've updated this (setting the style at the beginning of the function, but still allows for flexibility, as a different style could be specified for additional plots)

dgergel commented:
@jhamman ready for another review.

dgergel commented:
![36 6358_-99 5975](https://cloud.githubusercontent.com/assets/5017308/17536777/465f8eac-5e4c-11e6-977a-b2e222297210.png)

There is an issue with the observations for Net LW and SW, hence the 0s.

dgergel commented:
Monthly mean diurnal cycle, Fluxnet sites. Example figure for 36.6358, -99.5975.

![36 6358_-99 5975](https://cloud.githubusercontent.com/assets/5017308/17536829/de77b444-5e4c-11e6-90ba-dedf258469eb.png)

dgergel commented:
Daily SWE, Nov 2007 - July 2008, SNOTEL comparison. Example figure for 36.4875, -106.3130. 

Comparison figures also made for the following output variables (not shown): Albedo, latent heat, net radiation, sensible heat, snow canopy, snow depth, snow melt, and snow pack temperature. The additional output variable figures are just to compare VIC 4.2 and 5, as observations are not available. 

![36 4875_-106 3130](https://cloud.githubusercontent.com/assets/5017308/17536939/a3ca247a-5e4d-11e6-85ca-d304e79e6ed6.png)
